### PR TITLE
add alloydb module

### PR DESCRIPTION
**Pull request by prabhaarya**
Create Alloydb cluster and instance.

### CODE REVIEW

File: modules/alloydb-instance/variables.tf [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "automated_backup_policy" {
+  description = "The automated backup policy for this cluster. If no policy is provided then the default policy will be used. The default policy takes one backup a day, has a backup window of 1 hour, and retains backups for 14 days"
+  type = object({
+    location      = optional(string)
```
Comment (ludoo) [inline modules/alloydb-instance/variables.tf:20]: nit: I usually sort these with non-optional, simple type optional, complex type optional orderning

File: modules/alloydb-instance/variables.tf [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+}
+variable "cluster_id" {
+  description = "The ID of the alloydb cluster"
+  type        = string
+  validation {
+    condition     = can(regex("^[a-z0-9-]+$", var.cluster_id))
+    error_message = "ERROR: Cluster ID must contain only Letters(lowercase), number, and hyphen"
+  }
+}
+variable "cluster_display_name" {
```
Comment (ludoo) [inline modules/alloydb-instance/variables.tf:44]: I would remove the `cluster_` prefix from variable names, except for id

File: modules/alloydb-instance/variables.tf [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    password = string
+  })
+  default = null
+}
+variable "cluster_location" {
+  description = "Location where AlloyDb cluster will be deployed."
+  type        = string
+  # default     = "us-central1"
+}
+variable "primary_instance" {
```
Comment (ludoo) [inline modules/alloydb-instance/variables.tf:72]: nit: `primary_instance_config`

File: modules/alloydb-instance/variables.tf [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  validation {
+    condition     = can(regex("^[a-z]([a-z0-9-]{0,61}[a-z0-9])?$", var.primary_instance.instance_id))
+    error_message = "Primary Instance ID should satisfy the following pattern ^[a-z]([a-z0-9-]{0,61}[a-z0-9])?$"
+  }
+}
+variable "project_id" {
+  description = "The ID of the project in which to provision resources."
+  type        = string
+}
+variable "network_name" {
```
Comment (ludoo) [inline modules/alloydb-instance/variables.tf:97]: Do you need both name and self link?

File: modules/alloydb-instance/variables.tf [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  type    = string
+  default = "multiple-readpool"
+}
+variable "network_self_link" {
+  description = "Network ID where the AlloyDb cluster will be deployed."
+  type        = string
+}
+
```
Comment (ludoo) [inline modules/alloydb-instance/variables.tf]: `read_pool_instances`

File: tests/modules/alloydb_instance/examples/alloydb_instance.tfvars [hunk 6]
```diff
@@ -0,0 +1,40 @@
+project_id           = "myproject"
```
Comment (ludoo) [inline tests/modules/alloydb_instance/examples/alloydb_instance.tfvars:1]: can you move all examples to README examples?

File: modules/alloydb-instance/main.tf [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  read_pool_instance = (
```
Comment (ludoo) [inline modules/alloydb-instance/main.tf:18]: can you reorder locals alphabetically and break long lines? we have a few examples in our contributing guide

Comment (ludoo) [reply modules/alloydb-instance/main.tf:18]: `read_pool_instances`

File: modules/alloydb-instance/main.tf [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    content {
+      location      = automated_backup_policy.value.location
+      backup_window = automated_backup_policy.value.backup_window
+      enabled       = automated_backup_policy.value.enabled
+      labels        = automated_backup_policy.value.labels
+
+
+      weekly_schedule {
+        days_of_week = automated_backup_policy.value.weekly_schedule.days_of_week
+        dynamic "start_times" {
```
Comment (ludoo) [inline modules/alloydb-instance/main.tf:51]: if you use an `iterator` you can avoid long names

File: modules/alloydb-instance/main.tf [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+          for_each = { for i, time in automated_backup_policy.value.weekly_schedule.start_times : i => {
```
Comment (ludoo) [inline modules/alloydb-instance/main.tf:52]: can `weekly_schedule` be null?

File: modules/alloydb-instance/main.tf [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+  read_pool_config {
+    node_count = each.value.node_count
+  }
+
+  database_flags = each.value.database_flags
+  machine_config {
+    cpu_count = each.value.machine_cpu_count
+  }
+
+  depends_on = [google_alloydb_instance.primary, google_compute_network.default, google_compute_global_address.private_ip_alloc, google_service_networking_connection.vpc_connection]
```
Comment (ludoo) [inline modules/alloydb-instance/main.tf:143]: can you always break long lines?

File: modules/alloydb-instance/outputs.tf [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+output "cluster" {
+  description = "Cluster created"
+  value       = resource.google_alloydb_cluster.default.display_name
+  sensitive   = true
+}
+output "cluster_id" {
```
Comment (ludoo) [inline modules/alloydb-instance/outputs.tf:22]: missing newline here and below

**Review by ludoo — CHANGES_REQUESTED** [review]

File: modules/alloydb-instance/variables.tf (hunk 2 above)
Comment (prabhaarya) [reply modules/alloydb-instance/variables.tf:35]: Need to differentiate b/w cluster and primary instance in AlloyDB. So I kept `cluster_` prefix specifically for the cluster.

File: modules/alloydb-instance/README.md [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+## Simple example
+
+This example shows how to create Alloydb cluster and instance with multiple read pools in GCP project.
+
+```hcl
+module "alloydb" {
+  source               = "./fabric/modules/alloydb-instance"
+  project_id           = "myproject"
+  cluster_id           = "alloydb-cluster-all"
+  cluster_location     = "europe-west2"
```
Comment (sruffilli) [inline modules/alloydb-instance/README.md:15]: nit: should we use `region` instead to be consistant with the other modules?

Comment (ludoo) [reply modules/alloydb-instance/README.md:15]: if the resource attribute is location I would keep that, I seem to remember we do the same for GCS

**Review by ludoo — APPROVED** [review]

**Pull request merged**